### PR TITLE
feat: transformer tf text encoder

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
https://github.com/jina-ai/internal-tasks/issues/92

### CODE REVIEW

File: transformer_tf_text_encode.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            self.logger.warning('You tried to use a GPU but no GPU was found on'
+                                ' your system. Defaulting to CPU!')
+        tf.config.experimental.set_visible_devices(devices=cpus)
+
+    @requests
+    def encode(self, docs: DocumentArray, parameters: dict, *args, **kwargs):
+        """
+        Encode an array of string in size `B` into an ndarray in size `B x D`,
+        where `B` is the batch size and `D` is the dimensionality of the encoding.
+        :param docs: a document array of string type in size `B`
```
Comment (slettner) [inline transformer_tf_text_encode.py]:
```suggestion
        :param docs: DocumentArray containing images
        :param parameters: dictionary to define the `traversal_paths` and the `batch_size`. For example,
               `parameters={'traversal_paths': ['r'], 'batch_size': 10}`.
        :param kwargs: Additional key value arguments.
```

File: tests/unit/test_exec.py [hunk 2]
```diff
@@ -1,9 +1,56 @@
...
+    for path, count in [[['r'], 0], [['c'], 3], [['cc'], 0]]:
+        assert len(docs.traverse_flat(path).get_attributes('embedding')) == count
+        if count > 0:
+            assert docs.traverse_flat(path).get_attributes('embedding')[0].shape == (target_dim,)
+
+    encoder.encode(docs, parameters={'batch_size': 10, 'traversal_paths': ['cc']})
+    for path, count in [[['r'], 0], [['c'], 3], [['cc'], 2]]:
+        assert len(docs.traverse_flat(path).get_attributes('embedding')) == count
+        if count > 0:
+            assert docs.traverse_flat(path).get_attributes('embedding')[0].shape == (target_dim,)
```
Comment (slettner) [inline tests/unit/test_exec.py]: The `traversal_paths` could be parameterized and `traversal_paths=['r']` could be added.
Similar to here : https://github.com/jina-ai/executor-image-torch-encoder/blob/65fee77edf3d7758255c0ff38c39ad4e9693ece3/tests/integration/test_flow_integration.py#L48

File: tests/unit/test_exec.py [hunk 3]
```diff
@@ -1,9 +1,56 @@
...
+
+
+def test_no_documents():
+    encoder = TransformerTFTextEncoder()
+    docs = []
```
Comment (slettner) [inline tests/unit/test_exec.py]:
```suggestion
    docs = DocumentArray([])
```

File: tests/unit/test_exec.py [hunk 4]
```diff
@@ -1,9 +1,56 @@
...
+    encoder.encode(docs, parameters={'batch_size': 10, 'traversal_paths': ['r']})
+    assert not docs
```
Comment (slettner) [inline tests/unit/test_exec.py]: PEP 8: W292 no newline at end of file

File: README.md [hunk 5]
```diff
@@ -87,42 +65,44 @@ pods:
...
 
 ### Returns
 
-`Document` with `embedding` fields filled with an `ndarray` of the shape `embedding_dim` (=128, by default) with `dtype=nfloat32`.
-
-
-## 🔍️ Reference
-- Some reference
-
+`Document` with `embedding` fields filled with an `ndarray`  with `dtype=nfloat32`.
```
Comment (slettner) [inline README.md]:
```suggestion
`Document` with `embedding` fields filled with an `ndarray`  with `dtype=np.float32`.
```

File: README.md [hunk 6]
```diff
@@ -42,7 +20,7 @@ or in the `.yml` config.
 jtype: Flow
 pods:
   - name: encoder
-    uses: 'jinahub+docker://MyDummyExecutor'
+    uses: 'jinahub+docker://TransformerTFTextEncoder'
 ```
 
 #### using source codes
```
Comment (slettner) [inline README.md:26]: I would remove that section, looks very similar to the two snippets above.

**Review by slettner — CHANGES_REQUESTED** [review]
Nice work @Yongxuanzhang !

I left a couple of comments.

In general I think the tests could be extended like [here](https://github.com/jina-ai/executor-transformer-torch-encoder/tree/main/tests/unit). Those two executors are very similar, it should be mostly copy paste. But that is better in another PR.

Also I don't think the `__init__.py` files in the test directory are necessary, but that is minor.

**Comment by Yongxuanzhang** [comment]
> Nice work @Yongxuanzhang !
> 
> I left a couple of comments.
> 
> In general I think the tests could be extended like [here](https://github.com/jina-ai/executor-transformer-torch-encoder/tree/main/tests/unit). Those two executors are very similar, it should be mostly copy paste. But that is better in another PR.
> 
> Also I don't think the `__init__.py` files in the test directory are necessary, but that is minor.

Thanks for your comments. But removing  `__init__.py` will fail the CI

File: tests/unit/test_exec.py (hunk 2 above)
Comment (Yongxuanzhang) [reply tests/unit/test_exec.py]: how about now?

Comment (slettner) [reply tests/unit/test_exec.py]: Perfect, thanks!

**Review by slettner — APPROVED** [review]

**Pull request merged**